In [1]:
%matplotlib inline

# Problem Set 1

## Directions

1. Show all work/steps/calculations using a combination of code and Markdown. You will submit only this notebook named: [jhed_id].ipynb back to Blackboard from whence you downloaded it.
2. **All** work is to be your own. This is not a group project. You may, however, use code from the lectures and labs. Provide citations for any code that is not your own. You may also consult Stackoverflow, etc. This is not by any means "closed book" or anything like that. Basically, I need to see if *you* learned the concepts from the Lectures, *Fundamentals*, and Labs.
3. Add whatever markdown or code cells you need in each part to explain and calculate your answers. Don't just provide answers but explain them as well. **Explain and interpret your results.**

## Manipulating and Interpreting Probability

Given the following *joint probability distribution*, $P(A,B)$, for $A$ and $B$,


|        |    a1   |    a2   |

|----    |------|------|

| **b1** | 0.25 | 0.51 |

| **b2** | 0.19 | ?    |


Answer the following questions.

**1\. What is $P(A=a2, B=b2)$?**

In [2]:
print("P(A=a2,B=b2) = 1 - 0.25  - 0.51 - 0.19 = %.2f" %(1 - 0.25 - 0.51 - 0.19))

P(A=a2,B=b2) = 1 - 0.25  - 0.51 - 0.19 = 0.05


In [3]:
P = {"a1": {"b1": 0.25, "b2": 0.19}, "a2": {"b1": 0.51, "b2": 0.05}}

**2\. If I observe events from this probability distribution, what is the probability of seeing (a1, b1) then (a2, b2)?**

In [4]:
print("P((a1,b1)then(a2,b2)) = P(a1,b1) * P(a2,b2) = %.2f" %(P["a1"]["b1"]*P["a2"]["b2"]))

P((a1,b1)then(a2,b2)) = P(a1,b1) * P(a2,b2) = 0.01


**3\. Calculate the marginal probability distribution, $P(A)$.**

In [5]:
print("P(a1) = %.2f" % (P["a1"]["b1"] + P["a1"]["b2"]))
print("P(a2) = %.2f" % (P["a2"]["b1"] + P["a2"]["b2"]))

P(a1) = 0.44
P(a2) = 0.56


**4\. Calculate the marginal probability distribution, $P(B)$.**

In [6]:
print("P(b1) = %.2f" % (P["a1"]["b1"] + P["a2"]["b1"]))
print("P(b2) = %.2f" % (P["a1"]["b2"] + P["a2"]["b2"]))

P(b1) = 0.76
P(b2) = 0.24


**5\. Calculate the conditional probability distribution, $P(A|B)$.**

The formula for conditional probability is: P(A|B) = P(A,B)/P(B)

In [7]:
print( "P(a1|b1) = P(a1,b1)/P(b1) = %.2f" % ((P["a1"]["b1"])/(P["a1"]["b1"] + P["a2"]["b1"])))
print( "P(a2|b1) = P(a2,b1)/P(b1) = %.2f" % ((P["a2"]["b1"])/(P["a1"]["b1"] + P["a2"]["b1"])))
print( "P(a1|b2) = P(a1,b2)/P(b2) = %.2f" % ((P["a1"]["b2"])/(P["a1"]["b2"] + P["a2"]["b2"])))
print( "P(a2|b2) = P(a2,b2)/P(b2) = %.2f" % ((P["a2"]["b2"])/(P["a1"]["b2"] + P["a2"]["b2"])))

P(a1|b1) = P(a1,b1)/P(b1) = 0.33
P(a2|b1) = P(a2,b1)/P(b1) = 0.67
P(a1|b2) = P(a1,b2)/P(b2) = 0.79
P(a2|b2) = P(a2,b2)/P(b2) = 0.21


**6\. Calculate the conditional probability distribution, $P(B|A)$.**

The formula for conditional probability is: P(B|A) = P(B,A)/P(A)

In [8]:
print( "P(b1|a1) = P(b1,a1)/P(a1) = %.2f" % ((P["a1"]["b1"])/(P["a1"]["b1"] + P["a1"]["b2"])))
print( "P(b2|a1) = P(b2,a1)/P(a1) = %.2f" % ((P["a1"]["b2"])/(P["a1"]["b1"] + P["a1"]["b2"])))
print( "P(b1|a2) = P(b1,a2)/P(a2) = %.2f" % ((P["a2"]["b1"])/(P["a2"]["b1"] + P["a2"]["b2"])))
print( "P(b2|a2) = P(b2,a2)/P(a2) = %.2f" % ((P["a2"]["b2"])/(P["a2"]["b1"] + P["a2"]["b2"])))

P(b1|a1) = P(b1,a1)/P(a1) = 0.57
P(b2|a1) = P(b2,a1)/P(a1) = 0.43
P(b1|a2) = P(b1,a2)/P(a2) = 0.91
P(b2|a2) = P(b2,a2)/P(a2) = 0.09


**7\. Does $P(A|B) = P(B|A)$? What do we call the belief that these are always equal?**

The probably asks the various cases under P(A|B) = P(B|A)

If we take one case: P(a1|b1) and P(b1,a1)

Referring to answers to question 5 and 6 We can see:

In [9]:
print( "P(a1|b1) = P(a1,b1)/P(b1) = %.2f" % ((P["a1"]["b1"])/(P["a1"]["b1"] + P["a2"]["b1"])))
print( "P(b1|a1) = P(b1,a1)/P(a1) = %.2f" % ((P["a1"]["b1"])/(P["a1"]["b1"] + P["a1"]["b2"])))

P(a1|b1) = P(a1,b1)/P(b1) = 0.33
P(b1|a1) = P(b1,a1)/P(a1) = 0.57


Therefore, P(A|B) does not always equal to P(B|A)

The belief that P(A|B) and P(B|A) are always equal is called inverse probability fallacy

**8\. Does $P(A) = P(A|B)$? What does that mean about the independence of $A$ and $B$?**

No, P(A) does not equal to P(A|B). Referring to answers above:

In [10]:
print("P(a1) = %.2f" % (P["a1"]["b1"] + P["a1"]["b2"]))
print( "P(a1|b1) = %.2f" % ((P["a1"]["b1"])/(P["a1"]["b1"] + P["a2"]["b1"])))

P(a1) = 0.44
P(a1|b1) = 0.33


This means A and B are not independent

**9\. Using $P(A)$, $P(B|A)$, $P(B)$ from above, calculate,**

$P(A|B) = \frac{P(B|A)P(A)}{P(B)}$

Does it match your previous calculation for $P(A|B)$?

Yes, we can use P(a1|b1) to show the calculation as below:

In [11]:
print( "P(a1|b1) = (P(b1,a1)*P(a1))/P(b1) = %.2f" % ((((P["a1"]["b1"])/(P["a1"]["b1"] + P["a1"]["b2"])*(P["a1"]["b1"] + P["a1"]["b2"])))/(P["a1"]["b1"] + P["a2"]["b1"])))

P(a1|b1) = (P(b1,a1)*P(a1))/P(b1) = 0.33


## Question 1 - Bayes Rule

**Part 1.** There is a genetic condition that affects 2.3% of the population, C = {c, ~c}. Tests can detect the condition with 72.1% (true positive rate) but have a false positive rate of 20.3%, T={pos, neg}.
 
If we want to know `P(C|T)`, write out Bayes Rule for this problem:

$P(C|T) = \frac{P(T|C) * P(C)}{P(T)}$

Using the values supplied in the description of the problem, specify all priors and likelihoods for the version of Bayes Rule you just wrote out:

P(C) is the prior for the genetic condition and is described as (0.023, 0.977)
P(T|C) is the likelyhood of the condition detected given genetic condition
Based on above, probability given in the question, we have

In [12]:
print("P(c) = %.4f"%0.023)
print("P(~c) = %.4f"% (1-0.023))
print("P(t|c) = %.4f"%(0.721))
print("P(~t|c) = %.4f"%(1-0.721))
print("P(t|~c) = %.4f"%(0.203))
print("P(~t|~c) = %.4f"%(1-0.203))


P(c) = 0.0230
P(~c) = 0.9770
P(t|c) = 0.7210
P(~t|c) = 0.2790
P(t|~c) = 0.2030
P(~t|~c) = 0.7970


In [13]:
p1_prior = {"c": 0.023, "~c": 0.977}
p1_likelihoods = {"c": {"t": 0.721, "~t": 0.279}, "~c": {"t": 0.203, "~t": 0.797}}

Calculate `P(C|T)`. Please explain each individual probability and the difference between the prior and the posterior and the reason for the difference.

We can use the generic Bayes Rule code from Fundamentals and the lab for this calculation:

In [14]:
def normalize(dist):
    normalizer = sum( dist.values())
    for k in dist.keys():
        dist[ k] = dist[ k] / normalizer
    return dist # don't need to do this.

def query(prior, likelihoods, evidence):
    posterior = {}
    for k in prior.keys():
        posterior[ k] = likelihoods[ k][ evidence] * prior[ k]
    normalize( posterior)
    return posterior

Let's start with "t" condition detected as evidence

In [15]:
print(query(p1_prior, p1_likelihoods, "t"))

{'c': 0.07716109699693831, '~c': 0.9228389030030617}


`P(c|t)` = 0.0772 is the probability of having the genetic condition if the test detects the condition. This is higher than prior with the positive test result.

`P(~c|t)` = 0.9228 is the probability of not having the genetic condition if the test detects the condition. This is lower than prior with the positive test result.


Now let's use "~t" condition not detected as evidence

In [16]:
print(query(p1_prior, p1_likelihoods, "~t"))

{'c': 0.008173626838333635, '~c': 0.9918263731616663}


`P(c|~t)` = 0.0082 is the probability of having the genetic condition if the test does not detect the condition. This is lower than prior with the negative test result.

`P(~c|~t)` = 0.9918 is the probability of not having the genetic condition if the test does not detect the condition. This is higher than prior with the negative test result.

**Part 2.** Assume the true positive rate is 97.3% but the false positive rate is 37.2%. Recalculate P(C|T) and describe how the posterior distribution changes between Part 1 and Part 2 and *why*.



In [17]:
p2_prior = {"c": 0.023, "~c": 0.977}
p2_likelihoods = {"c": {"t": 0.973, "~t": 0.027}, "~c": {"t": 0.372, "~t": 0.628}}

Let's start with "t" condition detected as evidence

In [18]:
print(query(p2_prior, p2_likelihoods, "t"))

{'c': 0.058003281297382484, '~c': 0.9419967187026176}


`P(c|t)` = 0.0580 is the probability of having the genetic condition if the test detects the condition. This is lower than part 1 which means if someone gets a positive test result, they have a lower probability having the genetic condition.

`P(~c|t)` = 0.9420 is the probability of not having the genetic condition if the test detects the condition. This is higher than part 1 which means if someone gets a positive test result, they have a higher probability NOT having the genetic condition.


Now let's use "~t" condition not detected as evidence

In [19]:
print(query(p2_prior, p2_likelihoods, "~t"))

{'c': 0.001011109175368013, '~c': 0.998988890824632}


`P(c|~t)` = 0.0010 is the probability of having the genetic condition if the test does not detect the condition. This is lower than part 1 which means if someone gets a negative test result, they have a higher probability having the genetic condition.

`P(~c|~t)` = 0.9999 is the probability of not having the genetic condition if the test does not detect the condition. This is higher than part 1 which means if someone gets a negative test result, they have a higher probability NOT having the genetic condition.